Added KDF1 and KDF2 implementations

nickrmc83 opened the pull request:
KDF1 and KDF2 are not present in the  standard go crypto libraries. The
code includes Known Answer Tests and example usage.

Commit:
Added KDF1 and KDF2 implementations

KDF1 and KDF2 are not present in the  standard go crypto libraries. The
code includes Known Answer Tests and example usage.

Commit:
Adding go.mod file.

Commit:
Adding missing copyright headers

Added missing copyright headers

dmjones commented on README.md:
```diff
@@ -26,3 +26,9 @@ To build and test:
+
+KDF includes from-scratch implementations of the KDF1 and KDF2 key derivation algorithms used in various cryptographic schemes.
+The 2 flavors of KDF implement the standard `io.Reader` interface. The code includes known answer tests and example
+usage.
```
It would be nice to link to the RFCs.

nickrmc83 replied:
Added link in 281f9782a99f585e5feceef0e752111d15f140fd

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+	"crypto"
+	"errors"
+	"hash"
+	"io"
```
These imports are not sorted according to goimports. There is also an identation issue in the `var` block. File needs to be formatted.

nickrmc83 replied:
Fixed in 281f9782a99f585e5feceef0e752111d15f140fd

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+
+	// ErrInvalidSeedParameter a parameter is invalid.
+	ErrInvalidSeedParameter = errors.New("invalid input parameter")
+	)
```
Do we need to export these errors and commit to them being part of our API? I think we should un-export these, unless you can suggest some scenarios where someone's program might respond to the error.

nickrmc83 replied:
Removed errors from being exported

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+	)
+
+// Verify KDF completely implements the io.Reader interface.
+var _ io.Reader = &KDF{}
```
This is usually written `var _ io.Reader = (*KDF)(nil)`, if it's needed at all (since you have a `New...` function).

nickrmc83 replied:
Fixed in 281f9782a99f585e5feceef0e752111d15f140fd

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+// i2osp 4-byte integer marshalling.
+func i2osp(i uint32) (osp []byte) {
```
No need for a named return parameter.

nickrmc83 replied:
Done

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+	return
+}
+
+// NewKDF1 create a new KDF1 context.
```
The returned object is not safe to use from multiple go-routines. Should we mention that?

nickrmc83 replied:
```
The Go documentation follows a simple rule: If it is not explicitly stated that concurrent access to something is safe, it is not.
```
I think we're OK to follow that direction so we don't need to explicitly say these objects should not be accessed concurrently from multiple go-routines.

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+		return
+	}
+	// Calculate maximum size of the output based on the hash size.
+	var maxlen = int64(1<<32) * int64(hash.Size())
```
This probably makes sense if you read the RFC, but some extra inline comments explaining how you derive this maxlen would be nice.

nickrmc83 replied:
Added comment in 5b682ea71283039a49c814ee2900cccc4aa4b6ac as well as link to specification notes.

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+	return
+}
+
+func newKDF(seed, other []byte, hash crypto.Hash, offset uint32, length int) (kdf *KDF, err error) {
```
We don't need named return parameters.

nickrmc83 replied:
Changed in 281f9782a99f585e5feceef0e752111d15f140fd

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+// Read read the next len(p) bytes from the KDF context.
+func (kdf *KDF) Read(p []byte) (n int, err error) {
```
More named return parameters - we can live without these. Also, naked returns in larger functions get confusing.

nickrmc83 replied:
Fixed in 281f9782a99f585e5feceef0e752111d15f140fd

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+	toRead := min(len(p), kdf.length-kdf.position)
+	// When there's no data left fail.
+	if toRead == 0 {
+		err = io.EOF
```
Is this the right error if `p` is full and we haven't consumed all our KDF data?

nickrmc83 replied:
Fixed in 281f9782a99f585e5feceef0e752111d15f140fd. We only return `EOF` if there's no data left to consume from the KDF. Zero-length `p` will simply return `0, nil`.

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+		fromBuffer := min(len(kdf.buffer), toRead)
+		copy(p, kdf.buffer[:fromBuffer])
+		kdf.buffer = kdf.buffer[fromBuffer:]
+		n = fromBuffer
```
What happens if we've filled `p` with this buffered data? Shouldn't we exit here? I get that your logic below will catch this, but it feels cleaner to return early.

nickrmc83 replied:
Added early return clause in 9cc1e4096b7ce918fdb68b229c195fe819657418

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,149 @@
+		n = fromBuffer
+	}
+	// Calculate the number of full hash outputs required to satisfy request.
+	iterations := ((toRead - n) + (kdf.digester.Size() - 1)) / kdf.digester.Size()
```
Is this the same as `((toRead - n) / kdf.digester.Size()) + 1`? I find that easier to grok, but it may just be me.

Also, can we update `toRead` before we enter this part of the function. That will remove a lot of `toRead - n` statements that are hard to read.

nickrmc83 replied:
It is not the same. The code as written is equivalent to `math.Ceil(float64(toRead-n)/float64(kdf.digester.Size()))` I've broken this logic into a separate function to aid understanding and readability.

Commit:
Changes from code review

- Fixed import odering.
- Replaced named return variables with unnamed versions.
- Return EOF IFF there's no data left to read as opposed to a
zero-length buffer passed into `Read`

dmjones commented on kdf/kdf.go:
```diff
@@ -0,0 +1,148 @@
+	var n int
+	// When there's no data left return EOF.
+	if kdf.length-kdf.position == 0 {
+		return 0, io.EOF
```
If someone passes an empty slice for `p`, I think we should return `0, nil`, regardless of whether the KDF is empty. The `io.Reader` docs don't go as far as mandating this, but they say:

>Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0.

nickrmc83 replied:
This is now address in 9cc1e4096b7ce918fdb68b229c195fe819657418 whereby if `len(p)` is 0,  the function will return as if it would if it was satisfied from copying buffered data.

Commit:
Introduced `ceil` function for readability

- Added ceil function to aid readability.
- Fixed final digest block handling.

Commit:
Test case added to verify partial block read behaviour

Commit:
Return early if a `Read` is satisfied from cache

Add early return if a read is completely satisfied from cache or if
after the cached read there's no more data that can be returned.

Commit:
Added KDF maximum output size limiter comment

The maximum output size for a KDFx is specified in the specification and
is bounded by the use of a 32-bit unsigned counter for each digest
block.

dmjones commented:
LGTM

dmjones reviewed (approved):
LGTM

Commit:
Fixed maximum length calculation

The maximum input length calculation now validates the output will be
within accepted bounds for either KDF1 or KDF2 where previously this was
incorrect.

Commit:
Adding travisci config.

Commit:
Adding travis build status to README

Commit:
Removed extra `https://` from travis build status

The README linked to an invalid URL.

kaohsiungduck commented:
LGTM
